File: 490.迷宫.rs
/*
 * @lc app=leetcode.cn id=490 lang=rust
 *
 * [490] 迷宫
 */

// @lc code=start
impl Solution {
    pub fn has_path(maze: Vec<Vec<i32>>, start: Vec<i32>, destination: Vec<i32>) -> bool {
        use std::collections::HashSet;
        let (row, col) = (maze.len(), maze[0].len());
        let (mut stack, dirs, mut set) = (
            vec![start.clone()],
            vec![(-1, 0), (1, 0), (0, 1), (0, -1)],
            vec![format!("{}#{}", start[0], start[1])]
                .into_iter()
                .collect::<HashSet<String>>(),
        );
        while stack.len() > 0 {
            let block = stack.pop().unwrap();
            let (x, y) = (block[0], block[1]);
            if x == destination[0] && y == destination[1] {
                return true;
            }

            for d in dirs.iter() {
                let (mut i, mut j) = (x + d.0, y + d.1);
                while i >= 0
                    && j >= 0
                    && i < row as i32
                    && j < col as i32
                    && maze[i as usize][j as usize] == 0
                {
                    i += d.0;
                    j += d.1;
                }
                i -= d.0;
                j -= d.1;

                if !set.contains(&format!("{}#{}", i, j)) {
                    stack.push(vec![i, j]);
                    set.insert(format!("{}#{}", i, j).to_string());
                }
            }
        }
        false
    }
}
// @lc code=end

File: leetcode/490.迷宫.ts
/*
 * @lc app=leetcode.cn id=490 lang=typescript
 *
 * [490] 迷宫
 */

// @lc code=start
function hasPath(maze: number[][], start: number[], destination: number[]): boolean {
    const [stack, dirs, set] = [
        [start],
        [[-1, 0], [1, 0], [0, 1], [0, -1]],
        new Set([start.join("#")]),
    ];
    while (stack.length) {
        const [x, y] = stack.pop()!;
        if (x === destination[0] && y === destination[1]) return true;

        for (let [dx, dy] of dirs) {
            let [i, j] = [x + dx, y + dy];
            while (maze?.[i]?.[j] === 0) {
                i += dx;
                j += dy;
            }
            i -= dx, j -= dy;
            if (!set.has([i, j].join("#"))) {
                stack.push([i, j]);
                set.add([i, j].join("#"));
            }
        }
    }
    return false;
};
// @lc code=end

